doc/explanation/mental-model.rst

```diff
@@ -0,0 +1,175 @@
+   works, but should be a useful tool to build an understanding of Dune.
+
+How Dune Works
+--------------
+
+The building brick of Dune is the *rule*:
+
+    A *rule* reads *dependencies* and writes *targets* by executing an *action*.
+
+When ``dune build`` is executed, it will first read ``dune`` files to determine
+the rules that apply to the project. Once it has done this, it will determine
+what actions it needs to execute to build the required targets.
+
+An Example
+----------
+
+Let's take the following example.
+
+- there's a CLI tool written in OCaml.
+- it has some build-time configuration stored in ``config.json``.
+- it has an integration test which consists in running in it with
+  ``testdata.txt`` as input.
+
+Configuration Generation
+^^^^^^^^^^^^^^^^^^^^^^^^
+
+To express the generation of the configuration module we could write:
+
+.. code:: dune
+
+  (rule
+   (deps convert/json2ml.exe config.json)
+   (target config.ml)
+   (action
+    (run convert/json2ml.exe config.json -o config.ml)))
+
+This rule will:
+
+- read its dependencies: ``convert/json2ml.exe`` and ``config.json``
+- and write its target: ``config.ml``
+- using an action: ``(run convert/json2ml.exe config.json -o config.ml)``
+
+This rule is very explicit: we write a stanza for a single Dune rule.
+
+Building the Executable
+^^^^^^^^^^^^^^^^^^^^^^^
+
+In contrast, to describe the compilation of the executable, we would write:
+
+.. code:: dune
+
+  (executable
+   (name tool)
+   (modules main config))
+
+Here, we use Dune's abstractions. It knows about the OCaml compilation model:
+the modules need to be compiled and linked together. So it will generate the
+following rules under the hood:
+
+- one rule to compile the ``Main`` module:
+
+  - it will read its dependency: ``main.ml``
+  - and write its output: ``main.cmx``
+  - using an action: ``(run ocamlopt -c main.ml)``
+
+- one rule to compile the ``Config`` module:
+
+  - it will read its dependency: ``config.ml``
+  - and write its output: ``config.cmx``
+  - using an action: ``(run ocamlopt -c config.ml)``
+
+- one rule to compile the ``tool.exe`` executable:
+
+  - it will read its dependencies: ``main.cmx`` and ``config.cmx``
+  - and write its output: ``tool.exe``
+  - using an action: ``(run ocamlopt -o tool.exe main.cmx config.cmx``)
+
+Note that in this example, some files are targets of a rule and dependencies of
+another (``.cmx`` files). We are unlikely to ever interact with them directly,
+so it can also be useful to think of the ``(executable)`` stanza as acting as a
+big rule with ``main.ml`` and ``config.ml`` as inputs.
+
+Running the Tests
+-----------------
+
+Some rules do not produce any output file, but we're still interested in
+running an action. A test is a good example: we want the build process to exit
+with an error code if the action fails.
+In that case, the rule does not have targets, but we "attach" it to an
+:term:`alias`, ``runtest`` in this case.
```

Comment:
Maybe it's worth saying why we attach such actions to aliases: if we didn't, there would be no way to force executing them, since they have no targets. We could also remind the readers here that Dune is lazy: it only executes actions that are required to build the requested build goals.

Reply:
Agreed that it would be helpful to explain why the alias is necessary. My question reading this part is: Since dune doesn't run actions from rules whose targets all exist, and this rule has no targets, why does it ever run actions of rules with no targets? Is this a special case for rules with no targets?

Reply:
Doesn't it actually have some sort of hidden target? After all, if the test successfully finishes, it doesn't get re-run again.

I tripped up on this with e.g. property-based tests which don't have a reproducible outcome, once it succeeds you need `--force` to re-run it as there's no change in dependencies.

Reply:
I'm not sure about the details, but I think that the implementation of aliases is based on "tokens" (stamp files) whose name is based on the digest of their dependencies

Reply:
We shouldn't include such implementation details in the mental model. Stamp files are likely to disappear at some point (maybe even soon).

Reply:
(I'm thinking of nuking stamp files became writing a file on disk is a pretty inefficient way to cache an action execution.)